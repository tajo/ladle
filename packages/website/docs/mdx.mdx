---
id: mdx
title: MDX
---

Ladle supports MDX and markdown by default. You can use MDX to write your stories. This might be useful for documentation.

## Basic MDX Story

```md title="basic.stories.mdx"
# Header

> Some quote: Suspendisse at tempor velit. **Fusce** a fermentum arcu,
> vitae semper mi. Nunc placerat, mauris ac volutpat tempus, arcu eros
> accumsan nisi, in congue risus turpis in ligula.

Some [example](https://example.com) link.

1. One
2. Two
3. Three
```

This story will be displayed in the side nagivation as `Basic` > `Readme`.

## Custom Title

The title in side navigation can be customized:

```md title="basic.stories.mdx"
import { Meta } from "@ladle/react";

<Meta title="Documentation/Welcome" />

# Customized

This is a paragraph.
```

This story is displayed as `Documentation` > `Welcome`. If you set the `title` to `Welcome`, the story would be displayed as `Basic` > `Welcome`. You can use arbitrary number of levels.

## Multiple Stories

You can also create multiple stories within one `*.stories.mdx` file:

```md title="multiple.stories.mdx"
import { Story } from "@ladle/react";

This part will be rendered as a separate `Readme` story.

<Story name="First">
  <input />
  <button>simple</button>
</Story>

<Story name="Second">
  <input />
  <button>second</button>
</Story>
```

## Meta Parameters

You can also set `meta` parameters:

```md title="args.stories.mdx"
import { Story, Meta } from "@ladle/react";

<Meta meta={{ iframed: true }} />

This part will be rendered as a separate `Readme` story.

<Story name="First">
  <input />
  <button>simple</button>
</Story>

<Story name="Second" meta={{ width: 400 }}>
<input />
<button>second</button>
</Story>
```

## Importing Markdown

You can import markdown into your stories:

```md title="markdown.stories.mdx"
import Readme from "./README.md";

# Header

<Readme />
```
