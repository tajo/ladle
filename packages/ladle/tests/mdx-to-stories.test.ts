import { test, expect } from "vitest";
import { VFile } from "vfile";
import { createFormatAwareProcessors } from "@mdx-js/mdx/internal-create-format-aware-processors";
import mdxToStories from "../lib/cli/vite-plugin/mdx-to-stories.js";
import fs from "fs/promises";

const { process } = createFormatAwareProcessors({
  jsx: true,
});

const getFixture = async (path: string) => {
  const value = await fs.readFile(new URL(path, import.meta.url), {
    encoding: "utf8",
  });
  const file = new VFile({ value, path });
  const compiled = await process(file);
  return [String(compiled.value), path];
};

test("<Story /> component works", async () => {
  const fixture = await getFixture("./fixtures/story.stories.mdx");
  const output = await mdxToStories(...fixture);
  expect(output).toMatchSnapshot();
});

test("<Meta /> component works", async () => {
  const fixture = await getFixture("./fixtures/meta.stories.mdx");
  const output = await mdxToStories(...fixture);
  expect(output).toMatchSnapshot();
});
