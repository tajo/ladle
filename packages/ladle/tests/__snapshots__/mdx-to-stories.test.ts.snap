// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Meta /> component works 1`] = `
"/*@jsxRuntime automatic*/
/*@jsxImportSource react*/
import { Meta } from "@ladle/react";
import Readme from "./README.md";
function _createMdxContent(props) {
  const _components = {
    h1: "h1",
    ...props.components
  };
  return <><Meta title="Documentation" />{"\\n"}<_components.h1>{"Some Description"}</_components.h1></>;
}
export function MDXContent(props = {}) {
  const {
    wrapper: MDXLayout
  } = props.components || {};
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
MDXContent.storyName = "Documentation";"
`;

exports[`<Story /> component works 1`] = `
"/*@jsxRuntime automatic*/
/*@jsxImportSource react*/
import { Story } from "@ladle/react";
function _createMdxContent(props) {
  const _components = {
    code: "code",
    h1: "h1",
    p: "p",
    ...props.components
  };
  return <><_components.h1>{"MDX Button"}</_components.h1>{"\\n"}<_components.p>{"With "}<_components.code>{"MDX"}</_components.code>{", you can define button."}</_components.p>{"\\n"}<Story name="First"><input /><button>{"simple"}</button></Story>{"\\n"}<Story name="Second" args={{
      disabled: true
    }}><input /><button>{"second"}</button></Story></>;
}
export function MDXContent(props = {}) {
  const {
    wrapper: MDXLayout
  } = props.components || {};
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
export const LadleStory0 = () => <><input /><button>{"simple"}</button></>;
LadleStory0.storyName = "First";
export const LadleStory2 = () => <><input /><button>{"second"}</button></>;
LadleStory2.storyName = "Second";
LadleStory2.args = {
  disabled: true
};
MDXContent.storyName = 'Readme';"
`;
